keps/sig-api-machinery/20190325-unions.md

```diff
@@ -214,24 +214,70 @@ below.
+| A5       | `a: {...}`                       | `a: {...}`                       | `discriminator: A`<br>`a: {...}` |
+| A6       | `a: {...}`                       | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` |
+| A7       | `b: {...}`                       | empty                            | `discriminator: B`<br>`b: {...}` |
+| A8       | `b: {...}`                       | `a: {...}`                       | `discriminator: B`<br>`b: {...}` |
+| A9       | `b: {...}`                       | `discriminator: A`<br>`a: {...}` | `discriminator: B`<br>`b: {...}` |
+| A10      | `discriminator: A`<br>`a: {...}` | empty                            | `discriminator: A`<br>`a: {...}` |
+| A11      | `discriminator: A`<br>`a: {...}` | `a: {...}`                       | `discriminator: A`<br>`a: {...}` |
+| A12      | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` | `discriminator: A`<br>`a: {...}` |
+| A13      | `discriminator: B`<br>`b: {...}` | empty                            | `discriminator: B`<br>`b: {...}` |
+| A14      | `discriminator: B`<br>`b: {...}` | `a: {...}`                       | `discriminator: B`<br>`b: {...}` |
+| A15      | `discriminator: B`<br>`b: {...}` | `discriminator: A`<br>`a: {...}` | `discriminator: B`<br>`b: {...}` |
+
+Example updates with invalid content in the new object:
+
+| Scenario | New object                       | Existing object                  | Result                                                                      |
+|----------|----------------------------------|----------------------------------|-----------------------------------------------------------------------------|
+| B1       | `a: {...}, b: {...}`             | empty                            | fail<br>(two new fields set)                                                |
+| B2       | `a: {...}, b: {...}`             | `a: {...}`                       | fail<br>(two new fields set)                                                |
+| B3       | `a: {...}, b: {...}`             | `discriminator: A`<br>`a: {...}` | fail<br>(two fields set)                                                    |
+| B4       | `a: {...}, b: {...}`             | `discriminator: B`<br>`b: {...}` | fail<br>(two fields set)                                                    |
+| B5       | `discriminator: A`<br>`b: {...}` | empty                            | fail<br>(new field and changed discriminator don't match)                   |
+| B6       | `discriminator: A`<br>`b: {...}` | `a: {...}`                       | fail<br>(new field and changed discriminator don't match)                   |
+| B8       | `discriminator: B`<br>`a: {...}` | empty                            | fail<br>(new field and changed discriminator don't match)                   |
+| B7       | `discriminator: A`<br>`b: {...}` | `discriminator: A`<br>`a: {...}` | `discriminator: B`<br>`b:{...}`<br>(field replace, discriminator unchanged) |
```

Comment:
> are we planning to add explicit discriminators to existing types?

Yes

Reply:
can we limit this fixup to places where discriminators are added to existing types?

accepting an object with `discriminator: A` and turning it into `discriminator: B` is eyebrow-raising in general. I can see why it would be needed to keep (get deployment, clear volumeSource.foo, set volumeSource.bar, put deployment) working for existing clients, but I don't think we'd want that sort of error-masking behavior for all new discriminated union types going forward.

Reply:
We would either need a separate discriminator tag, or we could also base this behavior based on discriminator being optional/mandatory (though it would still have to be a different tag in the openapi). We should probably do both: `+optionalDiscriminator` (must be set on optional) and `+discriminator` must be required. `optionalDiscriminator` doesn't really sound great, but it would have to be a name bad enough that no-one is tempted to use it.

Reply:
> `optionalDiscriminator` doesn't really sound great, but it would have to be a name bad enough that no-one is tempted to use it.

Yeah, maybe a bad-sounding name like `+unionDiscriminatorAutodetectDeprecated` (since it's in service of fixup of existing API types, similar to `+unionDeprecated`). Per this example, it's not just missing discriminators we default, but present-and-incorrect ones we'd override based on the diff.

Reply:
> similar to `+unionDeprecated`

regarding that, I was confused by that name on first read... since the field it tags is not a union, and unions are not deprecated (and it's slightly weird to have both `+union` and `+unionDeprecated`). Something like `+unionMemberDeprecated` would have made more sense to me.